crypto/ocsp/ocsp_extension.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2015-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+// Modifications Copyright Amazon.com, Inc. or its affiliates.
+
+#include <openssl/mem.h>
+#include <openssl/rand.h>
+#include <string.h>
+
+#include "internal.h"
+
+#define OCSP_DEFAULT_NONCE_LENGTH 16
+
+int OCSP_REQUEST_get_ext_by_NID(OCSP_REQUEST *req, int nid, int lastpos) {
+  return X509v3_get_ext_by_NID(req->tbsRequest->requestExtensions, nid,
+                               lastpos);
+}
+
+X509_EXTENSION *OCSP_REQUEST_get_ext(OCSP_REQUEST *req, int loc) {
+  return X509v3_get_ext(req->tbsRequest->requestExtensions, loc);
+}
+
+int OCSP_BASICRESP_get_ext_by_NID(OCSP_BASICRESP *bs, int nid, int lastpos) {
+  return X509v3_get_ext_by_NID(bs->tbsResponseData->responseExtensions, nid,
+                               lastpos);
+}
+
+X509_EXTENSION *OCSP_BASICRESP_get_ext(OCSP_BASICRESP *bs, int loc) {
+  return X509v3_get_ext(bs->tbsResponseData->responseExtensions, loc);
+}
+
+static int ocsp_add_nonce(STACK_OF(X509_EXTENSION) **exts, unsigned char *val,
+                          int len) {
+  unsigned char *tmpval;
+  ASN1_OCTET_STRING os;
+  int ret = 0;
+  if (len <= 0) {
+    len = OCSP_DEFAULT_NONCE_LENGTH;
+  }
+  // Create the OCTET STRING manually by writing out the header and
+  // appending the content octets. This avoids an extra memory allocation
+  // operation in some cases. Applications should *NOT* do this because it
+  // relies on library internals.
+  os.length = ASN1_object_size(0, len, V_ASN1_OCTET_STRING);
+  if (os.length < 0) {
+    return 0;
+  }
+
+  os.data = OPENSSL_malloc(os.length);
+  if (os.data == NULL) {
+    goto err;
+  }
+  tmpval = os.data;
+  ASN1_put_object(&tmpval, 0, len, V_ASN1_OCTET_STRING, V_ASN1_UNIVERSAL);
+  if (val != NULL) {
+    memcpy(tmpval, val, len);
```

Comment:
This should use `OPENSSL_memcpy`. I also think this should be clarified in the documentation. What if someone passes in a 2 byte nonce and 0 for len, this would copy OCSP_DEFAULT_NONCE_LENGTH and read past the end of `val`. Is this a case where we should make our API a bit safer and require:
1. val == null and len can be 0 for default or a particular nonce size
2. val != null and len must be the length of val
Letting val !- null and picking a different len as default seems very dangerous.